renderdoc/driver/vulkan/vk_acceleration_structure.cpp

```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
+    // directly into upload memory
+    ser.Serialise("AS Input"_lit, contents, geomData.memSize, SerialiserFlags::NoFlags).Hidden();
 
-    uploadMemory = m_pDriver->AllocateMemoryForResource(true, mrq, MemoryScope::InitialContents,
-                                                        MemoryType::Upload);
-
-    if(uploadMemory.mem == VK_NULL_HANDLE)
-      return false;
+    if(!IsStructuredExporting(state) && uploadMemory.mem != VK_NULL_HANDLE)
+    {
+      // first ensure we flush the writes from the cpu to gpu memory
+      const VkMappedMemoryRange range = {
+          VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE,    //
+          NULL,
+          uploadMemory.mem,
+          0,
+          AlignUp(geomData.memSize, nonCoherentAtomSize),
+      };
 
-    vkr = m_pDriver->vkBindBufferMemory(d, uploadBuf, uploadMemory.mem, uploadMemory.offs);
-    CHECK_VKR(m_pDriver, vkr);
+      vkr = ObjDisp(d)->FlushMappedMemoryRanges(Unwrap(d), 1, &range);
+      CHECK_VKR(m_pDriver, vkr);
 
-    mappedMem = uploadMemory;
+      ObjDisp(d)->UnmapMemory(Unwrap(d), uploadMemory.mem);
+
+      // Allocate GPU memory and copy the AS input upload data into it
+      const VkBufferCreateInfo gpuBufInfo = {
+          VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+          NULL,
+          0,
+          geomData.memSize,
+          VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT |
+              VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT |
+              VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR,
+      };
 
-    vkr = ObjDisp(d)->MapMemory(Unwrap(d), Unwrap(mappedMem.mem), mappedMem.offs,
-                                AlignUp(mappedMem.size, nonCoherentAtomSize), 0, (void **)&contents);
-    CHECK_VKR(m_pDriver, vkr);
+      VkBuffer gpuBuf = VK_NULL_HANDLE;
+      vkr = m_pDriver->vkCreateBuffer(d, &gpuBufInfo, NULL, &gpuBuf);
+      CHECK_VKR(m_pDriver, vkr);
 
-    if(!contents)
-    {
-      RDCERR("Manually reporting failed memory map");
-      CHECK_VKR(m_pDriver, VK_ERROR_MEMORY_MAP_FAILED);
-      return false;
-    }
+      const MemoryAllocation gpuMemory = m_pDriver->AllocateMemoryForResource(
+          gpuBuf, MemoryScope::InitialContents, MemoryType::GPULocal);
+      if(gpuMemory.mem == VK_NULL_HANDLE)
+      {
+        RDCERR("Failed to allocate AS build data GPU buffer");
+        return false;
+      }
 
-    if(vkr != VK_SUCCESS)
-      return false;
-  }
+      vkr = ObjDisp(d)->BindBufferMemory(Unwrap(d), Unwrap(gpuBuf), Unwrap(gpuMemory.mem),
+                                         gpuMemory.offs);
+      CHECK_VKR(m_pDriver, vkr);
 
-  // not using SERIALISE_ELEMENT_ARRAY so we can deliberately avoid allocation - we serialise
-  // directly into upload memory
-  ser.Serialise("Serialised AS"_lit, contents, contentsSize, SerialiserFlags::NoFlags).Important();
+      VkCommandBuffer cmd = m_pDriver->GetInitStateCmd();
```

Comment:
`GetInitStateCmd()` is only intended to be used during Prepare and Apply, not during Serialise. I think this might work right now but that's somewhat coincidental and shouldn't be counted on. I think it's better to allocate the memory here and then do the copy with the build on apply (see the comment about caching the AS itself).